File: 2793-count-the-number-of-complete-components/count-the-number-of-complete-components.java
class Solution {
    HashMap<Integer,List<Integer>> mp;
    public int countCompleteComponents(int n, int[][] edges) {
        mp=new HashMap<>();
        for(int i=0;i<n;i++){
            mp.put(i,new ArrayList<>());
        }
        for(int[] e:edges){
            int v1=e[0];
            int v2=e[1];
            mp.get(v1).add(v2);
            mp.get(v2).add(v1);
        }
        return bft(n);
    }
    public int bft(int n){
        Queue<Integer> q=new LinkedList<>();
        HashSet<Integer> st=new HashSet<>();
        int ans=0;
        for(int src:mp.keySet()){
            if(st.contains(src)){
                continue;
            }
            List<Integer> ll=new ArrayList<>();
            q.add(src);
            int edge=0;
            while(!q.isEmpty()){
                // remove
                int r=q.poll();

                // ignore
                if(st.contains(r)){
                    continue;
                }

                // mark visited
                st.add(r);
                ll.add(r); //no of nodes in complete component

                // self work

                // add nbrs
                for(int nbrs:mp.get(r)){
                    edge++;//no of edges in complete component
                    if(!st.contains(nbrs)){
                        q.add(nbrs);
                    }
                }
            }

            // complete component condition
            int vtx=ll.size();
            if((edge/2) == vtx*(vtx-1)/2){
                ans++;
            }
        }
        return ans;
    }
}

File: 2793-count-the-number-of-complete-components/count-the-number-of-complete-components.cpp
class Solution {
public:
    int countCompleteComponents(int n, vector<vector<int>>& edges) {
        vector<vector<int>> ll(n);
        for(auto &e:edges){
            ll[e[0]].push_back(e[1]);
            ll[e[1]].push_back(e[0]);
        }
        return count(n,ll);
    }
    int count(int n, vector<vector<int>> &ll){
        queue<int> q;
        set<int> st;
        int c=0;
        for(int i=0;i<n;i++){
            if(st.count(i)){
                continue;
            }
            int v=0;
            int e=0;
            q.push(i);
            while(!q.empty()){
                int r=q.front();
                q.pop();
                if(st.count(r)){
                    continue;
                }
                v++;
                st.insert(r);
                for(int nbr:ll[r]){
                    e++;
                    if(!st.count(nbr)){
                        q.push(nbr);
                    }
                }
            }
            if(e/2==v*(v-1)/2){
                c++;
            }

        }
        return c;
    }
};